// @swc-node/core/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.transform = exports.transformJest = exports.transformSync = void 0;
const core_1 = require("@swc/core");
// Oldest LTS Node.js supported target
const DEFAULT_ES_TARGET = 'es2018';
function transformOption(path, options, jest = false) {
    var _a, _b, _c, _d, _e, _f;
    const opts = options == null ? {} : options;
    opts.esModuleInterop = (_a = opts.esModuleInterop) !== null && _a !== void 0 ? _a : true;
    return {
        filename: path,
        jsc: ((_b = options === null || options === void 0 ? void 0 : options.swc) === null || _b === void 0 ? void 0 : _b.swcrc)
            ? undefined
            : {
                target: (_c = opts.target) !== null && _c !== void 0 ? _c : DEFAULT_ES_TARGET,
                externalHelpers: jest ? true : Boolean(opts.externalHelpers),
                parser: {
                    syntax: 'typescript',
                    tsx: typeof opts.jsx !== 'undefined' ? opts.jsx : path.endsWith('.tsx'),
                    decorators: Boolean(opts.experimentalDecorators),
                    dynamicImport: Boolean(opts.dynamicImport),
                },
                transform: {
                    legacyDecorator: Boolean(opts.experimentalDecorators),
                    decoratorMetadata: Boolean(opts.emitDecoratorMetadata),
                    react: options === null || options === void 0 ? void 0 : options.react,
                    // @ts-expect-error
                    hidden: {
                        jest,
                    },
                },
                keepClassNames: opts.keepClassNames,
                paths: opts.paths,
                baseUrl: opts.baseUrl,
            },
        minify: false,
        isModule: true,
        module: ((_d = options === null || options === void 0 ? void 0 : options.swc) === null || _d === void 0 ? void 0 : _d.swcrc)
            ? undefined
            : {
                type: (_e = options === null || options === void 0 ? void 0 : options.module) !== null && _e !== void 0 ? _e : 'commonjs',
                noInterop: !opts.esModuleInterop,
            },
        sourceMaps: jest || typeof opts.sourcemap === 'undefined' ? 'inline' : opts.sourcemap,
        inlineSourcesContent: true,
        swcrc: false,
        ...((_f = options === null || options === void 0 ? void 0 : options.swc) !== null && _f !== void 0 ? _f : {}),
    };
}
function transformSync(source, path, options) {
    return (0, core_1.transformSync)(source, transformOption(path, options));
}
exports.transformSync = transformSync;
function transformJest(source, path, options) {
    return (0, core_1.transformSync)(source, transformOption(path, options, true));
}
exports.transformJest = transformJest;
function transform(source, path, options) {
    return (0, core_1.transform)(source, transformOption(path, options));
}
exports.transform = transform;
